consensus/src/lib.rs

```diff
@@ -26,28 +52,135 @@ pub enum ProgressResult {
     NonNilPreComitted(ConsensusRound, Hash256, Timestamp),
     NilPreVoted(ConsensusRound, Timestamp),
     NilPreComitted(ConsensusRound, Timestamp),
-    Finalized(Timestamp),
+    Finalized(Hash256, Timestamp),
+}
+
+pub struct ConsensusMessageFilter {
+    /// Note that it is even DESIRABLE to use a synchronous lock in the async context,
+    /// if it is guaranteed that the lock is not held for a long time.
+    verified_block_hash: Arc<parking_lot::RwLock<BTreeSet<Hash256>>>,
+    validator_set: BTreeSet<PublicKey>,
+}
+
+impl MessageFilter for ConsensusMessageFilter {
+    fn filter(&self, message: &Message) -> Result<(), String> {
+        serde_json::from_str::<ConsensusMessage>(message.data()).map_err(|e| e.to_string())?;
+        if !self.validator_set.contains(message.signature().signer()) {
+            return Err("The signer is not in the validator set".to_string());
+        }
+        if self
+            .verified_block_hash
+            .read()
+            .contains(&message.to_hash256())
+        {
+            Ok(())
+        } else {
+            Err("The block hash is not validated yet.".to_string())
+        }
+    }
 }
 
 pub struct Consensus<N: GossipNetwork, S: Storage> {
-    pub dms: DMS<N, S>,
+    dms: DMS<N, S>,
+    /// A cache of the consensus state.
+    state: State,
+    /// The set of the block hashes that have been verified, shared by the message filter.
+    ///
+    /// Note that there is the exactly same copy in the `state`.
+    verified_block_hash: Arc<parking_lot::RwLock<BTreeSet<Hash256>>>,
+    /// (If participated) the private key of this node
+    this_node_key: Option<PrivateKey>,
 }
 
 impl<N: GossipNetwork, S: Storage> Consensus<N, S> {
     pub async fn create(
-        _dms: DMS<N, S>,
-        _height: BlockHeight,
-        _validator_set: &[(PublicKey, VotingPower)],
+        dms: DMS<N, S>,
+        validator_set: &[(PublicKey, VotingPower)],
+        this_node_key: Option<usize>,
+        timestamp: Timestamp,
+        consensus_params: ConsensusParams,
     ) -> Result<(), Error> {
-        unimplemented!()
+        let this_node_index = this_node_key
+            .map(|key| {
+                validator_set
+                    .iter()
+                    .position(|(pk, _)| *pk == validator_set[key].0)
+                    .ok_or_else(|| anyhow::anyhow!("The validator set does not contain this node."))
+            })
+            .transpose()?;
+        let height_info = HeightInfo {
+            validators: validator_set.iter().map(|(_, v)| *v).collect(),
```

Comment:
why is it named validators? the proper meaning of the code is total voting power i guess.